internal/publish/serverservice_components.go

```diff
@@ -881,3 +911,42 @@ func (h *serverServicePublisher) setVersionedAttributes(component *serverservice
 		},
 	)
 }
+
+// addFirmwareData queries ServerService for the firmware version and try to find a match.
+func (h *serverServicePublisher) addFirmwareData(ctx context.Context, deviceVendor string, component *serverservice.ServerComponent, vattr *versionedAttributes) (vatrr *versionedAttributes, err error) {
+	// Filter by firmware vendor + version
+	// TODO: fix serverservice to accept component in filter
+	params := serverservice.ComponentFirmwareVersionListParams{
+		Vendor:  component.Vendor,
+		Version: vattr.Firmware.Installed,
+	}
+
+	// Query ServerService for the firmware
+	firmwares, _, err := h.client.ListServerComponentFirmware(ctx, &params)
```

Comment:
I suggest this data is cached from serverservice when the publisher is spawned [here](https://github.com/metal-toolbox/alloy/blob/FS-448-firmware-vendor/internal/publish/serverservice.go#L120) so that these queries aren't performed per component. On each run, Alloy run will re-spawn the publisher and so the data would be refreshed.

Reply:
I'd rather not cache this data. A few reasons for not doing so: 
- The `component_firmware_version` table is going to grow over time and is not as static as the `server_component_type` which may introduce bugs due to cache being stale;
- I'd rather find the limitation of ServerService with our internal tooling than optimise now and found out later that ServerService can't handle the amount of requests that we throw at it;
- We can always optimise as suggested later if it becomes an issue.

Reply:
These are my thoughts on the points you've mentioned,

> The component_firmware_version table is going to grow over time and is not as static as the server_component_type which may introduce bugs due to cache being stale;

The cache lifetime would be the length of each Alloy run - which is not more than an hour for an out of band collection of ~1000 servers and less than 5 secs for an inband collection. So I really don't think this would be a problem, especially since the next collection would have its cache refreshed.

> I'd rather find the limitation of ServerService with our internal tooling than optimise now and found out later that ServerService can't handle the amount of requests that we throw at it;

Theres other ways to find the limitations in serverservice (with a benchmarking utility), I'd rather not have Alloy take it down in an uncontrolled manner. We can perform a controlled chaos monkey drill in one facility that having it break the service (lets take up this operational topic outside of this repo).

> We can always optimise as suggested later if it becomes an issue.

What mainly concerns me here is that, Alloy now does the `number of component` times the request per server - for the same firmware data. And in the worst case - twice the count if the component vendor is not known. I do think the optimisation is worth the time and effort in this case.

Reply:
I don't really agree with this rationale and I think it's an anti pattern. Nevertheless  0bb2302c4719d8a436cf627fe0cd7bf314df949e adds a caching mechanism for the firmware data.